Add Garmin Connect integration

cyberjunky opened the pull request:
## Description:
Provides sensors for quite a few values exposed in Garmin Connect.

Implemented async and configflow, dropped yaml cfg for now

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#11752

## Example entry for `configuration.yaml` (if applicable):
```yaml
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

cyberjunky committed:
Added code files

cyberjunky committed:
Added code files

cyberjunky committed:
Correctly name init file

cyberjunky committed:
Correctly name init file

cyberjunky committed:
Update codeowners

cyberjunky committed:
Update codeowners

cyberjunky committed:
Update requirements

cyberjunky committed:
Update requirements

cyberjunky committed:
Black changes, added to coveragerc

cyberjunky committed:
Merge branch 'garmin-connect' of https://github.com/cyberjunky/home-assistant into garmin-connect

cyberjunky committed:
Removed documentation location for now

cyberjunky committed:
Added documentation url

cyberjunky committed:
Fixed merge

cyberjunky commented:
Looking into flake8 failure.

cyberjunky commented:
Need to decide about correct/wanted config name for the resources.
I'm now using the home-assistant's native in the code:

CONF_MONITORED_CONDITIONS

But FitBit uses a local declared:

CONF_MONITORED_RESOURCES = "monitored_resources"

And I believe there is also a:

CONF_RESOURCES

Any suggestion/preferences?

cyberjunky committed:
Fixed flake8 syntax

cyberjunky committed:
Merge branch 'garmin-connect' of https://github.com/cyberjunky/home-assistant into garmin-connect

cyberjunky committed:
Fixed isort

cyberjunky commented:
> Need to decide about correct/wanted config name for the resources.
> I'm now using the home-assistant's native in the code:
> 
> CONF_MONITORED_CONDITIONS

> Any suggestion/preferences?

Doing a fgrep -R inside the components directory shows that the usage of CONF_MONITORED_CONDITIONS is the highest, so sticking with that one.

balloob commented:
We're phasing out the usage of monitored conditions per [ADR-003](https://github.com/home-assistant/architecture/blob/master/adr/0003-monitor-condition-and-data-selectors.md). Please remove that support.

balloob commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+    "netRemainingKilocalories": ["Net Remaining KiloCalories", "kcal", "mdi:food"],
+    "netCalorieGoal": ["Net Calorie Goal", "cal", "mdi:food"],
+    "totalDistanceMeters": ["Total Distance Mtr", "mtr", "mdi:walk"],
+    "wellnessStartTimeLocal": ["Wellness Start Time", "", "mdi:clock"],
```
Please add support for device classes. For example, this should be `timestamp`. See https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

cyberjunky replied:
Done

balloob commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+    try:
+        garmin_client = garminconnect.Garmin(email, password)
+    except ValueError as err:
+        _LOGGER.error("Error occured during Garmin Connect Client init: %s", err)
```
Are you able to differentiate between bad credentials and no internet connection ? If so, for no internet you can `raise PlatformNotReady`

cyberjunky replied:
Done

balloob commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+    prefix_name = config.get(CONF_NAME)
+
+    try:
+        garmin_client = garminconnect.Garmin(email, password)
```
Do they have OAuth2 available ?

cyberjunky replied:
This integration is using the JSON rest API calls as used by the Garmin Connect website, and doesn't support OAuth as far as I could see.
The official Garmin API does, but I believe its a paid service. (I think with a $5000 fee)
I will have another look.

balloob commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+        """Fetch the latest data."""
+        today = date.today()
+        try:
+            self.data = self.client.fetch_stats(today.strftime("%Y-%m-%d"))
```
```suggestion
            self.data = self.client.fetch_stats(today.isoformat())
```

balloob commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+            }
+        return attributes
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
```
Why throttle this, `self.-data.update` is already throttled?

cyberjunky replied:
Fixed it

balloob commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+        self._data.update()
+        data = self._data.data
+
+        if GARMIN_CONDITIONS_LIST[self._type] and self._type in data:
```
What are you checking with `GARMIN_CONDITIONS_LIST[self._type]` ? In Python if a key doesn't exist and you use square brackets, it raises `KeyError`. How could it happen that `self._type` is changing and no longer in `GARMIN_CONDITIONS_LIST` ?

cyberjunky replied:
self._data.data contains the complete JSON data as returned by Garmin, it could contain a new  type in the future which the integration doesn't know off.
But the check is wrong as you said, I will revise this, maybe use a default type if unknown.

cyberjunky replied:
Never mind, I removed it, the type couldn't be wrong if the sensor was created with it.

cyberjunky replied:
I removed the checks.

balloob reviewed (approved):
Overall looks great. Few small changes.

I would suggest you add a config flow, it's really simple using `python3 -m script.scaffold config_flow` and it unlocks a lot of cool things like device info and device automations.

balloob commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+class GarminConnectSensor(Entity):
+    """Representation of a Garmin Connect Sensor."""
+
+    def __init__(self, data, sensor_type, name, unit, icon):
```
Do you have data to provide a unique ID? Maybe something as a device ID + sensor_type would be unique ?

cyberjunky replied:
I use users full name as filled in in Garmin Connect profile.

cyberjunky commented:
Thanks for the work you give me. :wink: No really, for me it's also about learning more/better Python.
Will try to implement all suggestions.

BTW: I noticed that the OTGW and Ring integrations -among others- dropped support to be configure 100% via yaml anymore (and force users to enable/disable entities via GUI) When you have integrations with a lot of sensors (OTGW >40, this one >70) this is a pain, for config and restore/re-configuring (which ones did I had enabled?)

At the State of the Union I heard that yaml configuration would always be kept possible?
Especially for backup and documentation purpose I find this important, I can understand this route for security reasons (no unencrypted credentials in secrets.yaml)
Maybe it's an idea to code some kind of Veracrypt type of vault in Python to store these safely?

MartinHjelmare commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+        "%",
+        "mdi:image-filter-hdr",
+    ],
+    "durationInMilliseconds": ["Duration", "ms", "mdi:progress-clock"],
```
What duration is this?

cyberjunky replied:
Unknown yet, I removed it as devicetype for now.

cyberjunky committed:
Removed false check and double throttle, applied time format change

cyberjunky committed:
Renamed email to username, used dict, deleted unused type, changed attr name

cyberjunky commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Required(CONF_EMAIL): cv.string,
```
Fixed this

cyberjunky commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Garmin Connect component."""
+
+    email = config.get(CONF_EMAIL)
```
Fixed this

cyberjunky commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+
+    email = config.get(CONF_EMAIL)
+    password = config.get(CONF_PASSWORD)
+    prefix_name = config.get(CONF_NAME)
```
Fixed this

cyberjunky commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+    entities = []
+    for resource in config[CONF_MONITORED_CONDITIONS]:
+        sensor_type = resource
+        name = prefix_name + " " + GARMIN_CONDITIONS_LIST[resource][0]
```
Implemented f-strings

cyberjunky commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,255 @@
+        if self._data.data:
+            attributes = {
+                "source": self._data.data["source"],
+                "last synced (GMT)": self._data.data["lastSyncTimestampGMT"],
```
I have removed (GMT) and made it using snake_case.

cyberjunky committed:
Async and ConfigFlow code

cyberjunky committed:
Fixes

balloob commented:
PRs are not the right venue to discuss architecture issues. 

The frontend makes it very easy to enable/disable entities with the new multi-select feature in 104.

Ring did not drop YAML. It just added support for config flows which now make it possible to support 2FA and not store the username/password in your configuration **at all**. It's a big win for security.

cyberjunky committed:
Added device_class and misc fixes

cyberjunky committed:
isort and pylint fixes

cyberjunky committed:
Removed from test requirements

cyberjunky committed:
Merge pull request #1 from cyberjunky/garmin-connect-configflow

Garmin connect configflow

cyberjunky commented:
Need to work on test code, that's new for me.

cyberjunky committed:
Fixed isort checkblack

cyberjunky committed:
Removed host field

MartinHjelmare commented on .coveragerc:
```diff
@@ -255,6 +255,9 @@ omit =
     homeassistant/components/futurenow/light.py
     homeassistant/components/garadget/cover.py
+    homeassistant/components/garmin_connect/__init__.py
+    homeassistant/components/garmin_connect/consts.py
```
const.py

cyberjunky replied:
Good catch, fixed.

cyberjunky committed:
Fixed coveragerc

cyberjunky committed:
Start working test file

cyberjunky commented:
I only need to finish test_config_flow.py code, but I'm stuck, no idea how to mock below part from config_flow.py (especially the _get_fullname() routine to mock returning a name string) from within test_config_flow.py
```
    async def async_step_user(self, user_input=None):
        """Handle the initial step."""
        if user_input is None:
            return await self._show_setup_form(user_input)

        errors = {}
        try:
            unique_id = await self._get_fullname(
                user_input[CONF_USERNAME], user_input[CONF_PASSWORD],
            )
        except GarminConnectConnectionError:
            errors["base"] = "cannot_connect"
            return await self._show_setup_form(errors)
        except GarminConnectAuthenticationError:
            errors["base"] = "invalid_auth"
            return await self._show_setup_form(errors)
        except GarminConnectTooManyRequestsError:
            errors["base"] = "too_many_requests"
            return await self._show_setup_form(errors)
        except Exception:  # pylint: disable=broad-except
            _LOGGER.exception("Unexpected exception")
            errors["base"] = "unknown"
            return await self._show_setup_form(errors)

        entries = self._async_current_entries()
        for entry in entries:
            if entry.data[CONF_ID] == unique_id:
                return self.async_abort(reason="already_setup")

        return self.async_create_entry(
            title=unique_id,
            data={
                CONF_ID: unique_id,
                CONF_USERNAME: user_input[CONF_USERNAME],
                CONF_PASSWORD: user_input[CONF_PASSWORD],
            },
        )

    async def _get_fullname(self, username: str, password: str) -> str:
        """Get full name from Garmin Connect profile."""
        return Garmin(username, password).get_full_name()
```
Help! :disappointed:

MartinHjelmare commented:
Patch the `Garmin` client class. You can set a specific return value of the patched class or return values of the instance methods, or set a side_effect to raise an exception.

https://github.com/home-assistant/home-assistant/blob/e8e7e66e2bcd18e61ba0c84f47c05fa974dfd640/tests/components/hue/test_config_flow.py#L375-L377

https://github.com/home-assistant/home-assistant/blob/e8e7e66e2bcd18e61ba0c84f47c05fa974dfd640/tests/components/brother/test_config_flow.py#L92

cyberjunky commented:
@MartinHjelmare thanks! I was trying to patch the stuff inside config_flow.py itself which I didn't get to work, now I understand why it still didn't get mocked, and tried to connect to real website.

MartinHjelmare commented:
Here are the docs:
https://docs.python.org/3/library/unittest.mock.html#where-to-patch

cyberjunky commented:
Arg..
Got this working, but have a pylint
Redefining name 'mock_garminconnect' from outer scope (line 28)

```

def init_config_flow(hass):
    """Init a configuration flow."""
    flow = config_flow.GarminConnectConfigFlowHandler()
    flow.hass = hass
    return flow


async def test_show_form(hass):
    """Test that the form is served with no input."""
    flow = init_config_flow(hass)

    result = await flow.async_step_user(user_input=None)

    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
    assert result["step_id"] == "user"


@pytest.fixture
def mock_garminconnect():
    """Mock garmin."""
    with patch("homeassistant.components.garmin_connect.config_flow.Garmin") as garmin:
        garmin().get_full_name.return_value = "First Lastname"
        yield garmin


async def test_step_user(hass, mock_garminconnect):
    """Test registering an integration and finishing flow works."""
    conf = {
        CONF_ID: "First Lastname",
        CONF_USERNAME: "my@email.address",
        CONF_PASSWORD: "mypassw0rd",
    }

    flow = init_config_flow(hass)
    result = await flow.async_step_user(user_input=conf)

    assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
    assert result["title"] == "First Lastname"
    assert result["data"] == {
        CONF_ID: "First Lastname",
        CONF_USERNAME: "my@email.address",
        CONF_PASSWORD: "mypassw0rd",
    }
```

MartinHjelmare commented:
Set a name of the fixture and rename the fixture function.

```py
@pytest.fixture(name="mock_garmin_connect")
def mock_garmin_connect_fixture():
    ...
```

cyberjunky committed:
Added more config_flow tests

cyberjunky committed:
Enable only most used sensors by default

cyberjunky committed:
Added more default enabled sensors, fixed tests

cyberjunky committed:
Fixed isort

cyberjunky commented:
I removed the WIP title flag, I think my code is ready and finished, applied all requests as far as portal supported them, updated documentation in home-assistant.io PR
Not sure about the failing CI test though, cannot find a relation to my code?

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,122 @@
+    """Test that the form is served with no input."""
+    flow = init_config_flow(hass)
+
+    result = await flow.async_step_user(user_input=None)
```
Start the flow like this without user input:
https://github.com/home-assistant/home-assistant/blob/80a55360dc0cba19f2b7ccfdc84ea369ae368f73/tests/components/hue/test_config_flow.py#L140-L142

cyberjunky replied:
This works, adapted.

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,122 @@
+    """Test registering an integration and finishing flow works."""
+    flow = init_config_flow(hass)
+
+    result = await flow.async_step_user(user_input=MOCK_CONF)
```
Start the flow with user input like this:
https://github.com/home-assistant/home-assistant/blob/80a55360dc0cba19f2b7ccfdc84ea369ae368f73/tests/components/hue/test_config_flow.py#L395-L397

The example above starts the import step. Replace that string as needed, with eg "user".

cyberjunky replied:
This doesn't run, because:
`homeassistant.data_entry_flow.UnknownStep: Handler GarminConnectConfigFlowHandler doesn't support step import`
I didn't create it because I only have user input from integration config, no yaml based config, or is that wrong assumption?

cyberjunky replied:
Wrote it like this:
```
    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": "user"},
        data={"username": "my@email.address", "password": "mypassw0rd"},
    )
```

MartinHjelmare replied:
Yes, that's why I wrote that you replace the import string as needed with eg "user".

cyberjunky replied:
You are right, overlooked it.

MartinHjelmare reviewed (commented):
Please use the core interfaces in the test to make them more robust.

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,122 @@
+    flow = init_config_flow(hass)
+
+    with patch(
+        "homeassistant.components.garmin_connect.config_flow.Garmin",
```
Is it the instantiation of this class that will raise or is it when the method of the class instance is called? Ie, does the real class instantiation have real side effects?

If it's a method that should raise we could simplify and use the mocked instance that the fixture yields and just set a side_effect on the corresponding instance method.

cyberjunky replied:
I used:
`"homeassistant.components.garmin_connect.config_flow.Garmin",`
because it mocked two methods at ones. I couldn't get them to get mocked directly like I think you prefer. (not using *config_flow* but garmin directly)
this is the code called from my configflowhandler:
```
client = garminconnect.Garmin(username, password)
client.get_full_name()
```
Cannot get to patch both instead of 
`"homeassistant.components.garmin_connect.config_flow.Garmin",`

MartinHjelmare replied:
We would keep the fixture with the change I suggested above. This already patches the `Garmin` class. Then we can set return values or side effects on the class instance as needed here in the test. Ie, we don't need to patch again here.

MartinHjelmare replied:
Eg:
```py
mock_garmin_connect.get_full_name.side_effect = Exception()
```

cyberjunky replied:
Clear... working on that...thanks for the patience :smile:

cyberjunky commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,122 @@
+        yield garmin
+
+
+def init_config_flow(hass):
```
Done

cyberjunky committed:
Test config_flow  improvements

cyberjunky committed:
Remove unused import

cyberjunky committed:
Removed redundant patch calls

MartinHjelmare commented on homeassistant/components/garmin_connect/__init__.py:
```diff
@@ -0,0 +1,104 @@
+    password = entry.data[CONF_PASSWORD]
+
+    try:
+        garmin_client = Garmin(username, password)
```
It's a better pattern to not have the class instantiation do side effects. Instead have the library user call `garmin_client.login()` to log in. Actions should be explicit.

cyberjunky replied:
The communication flow with the portal is not standard, the Garmin init class does a get_full_name to get a displaystring (an id to be used to fetch data), so the login call is only called first time, and when session expires, the rest of the communication is done using a single get_stats().
Even when you provide wrong credentials, the portal also return status code 200, (not unauth) but the returned html doesn't contain the data needed.

For example the Toomanyrequests exception can occur at each call, so I implemented all exceptions on all calls.
I will try to rewrite the library so we call .login() instead if the Garmin class itself.
And use same or less api calls.

cyberjunky replied:
I have now separated the Garmin.login method from the Class init, and fixed the exception catching. If anything else let me know.
Btw it would be nice to be able to also link integration/sensors to a Person instead of an Area using the GUI.

cyberjunky replied:
But now the test_config_flows fail again, this test code has cost me double the time it took to wrote the whole integration includeing the python module. Very frustrating.

cyberjunky committed:
Fixed mock return value

paularmstrong commented:
Just want to poke in and say thanks @cyberjunky for diving into this and @MartinHjelmare for all of your patience and assistance on the code review.

Really looking forward to having this integration for myself!

cyberjunky committed:
Updated to garmin_connect 0.1.8 fixed exceptions

cyberjunky committed:
Quick fix test patch to see if rest is error free

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,122 @@
+    """Mock Garmin."""
+    with patch("homeassistant.components.garmin_connect.config_flow.Garmin") as garmin:
+        garmin().get_full_name.return_value = "First Lastname"
+        yield garmin
```
Bump.

cyberjunky replied:
Got it.

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+    """Test for connection error."""
+    with patch(
+        "homeassistant.components.garmin_connect.config_flow.Garmin.login",
+        side_effect=GarminConnectConnectionError("errormsg"),
```
Remove the patch and do this instead.
```py
mock_garmin_connect.login.side_effect = GarminConnectConnectionError("errormsg")
```

cyberjunky replied:
I tried that, but then the mock fails, it gets the website.
Will have another try/look. So close...

MartinHjelmare replied:
Did you change the fixture yet? I bumped my comment above.

cyberjunky replied:
Change and seem to have fixed it, thanks a lot!
This integration is now equally written then most in core as far as I can see... 
Who determines Quality Scales? :smile:
I guess the CI test still fails because of an issue in another parts of the branch.

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+
+async def test_authentication_error(hass):
+    """Test for authentication error."""
+    with patch(
```
See above.

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+
+async def test_toomanyrequest_error(hass):
+    """Test for toomanyrequests error."""
+    with patch(
```
See above.

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+
+async def test_unknown_error(hass):
+    """Test for unknown error."""
+    with patch(
```
See above.

cyberjunky committed:
Fixed mock routine

cyberjunky commented:
I see this error in the CI logs pytest for python37:
Is this something I can/need to fix?
```
import file mismatch:
imported module 'test_config_flow' has this __file__ attribute:
  /__w/1/s/tests/components/garmin_connect/test_config_flow.py
which is not the same as the test file we want to collect:
  /__w/1/s/tests/components/netatmo/test_config_flow.py
HINT: remove __pycache__ / .pyc files and/or use a unique basename for your test file modules
```

MartinHjelmare commented:
The netatmo integration is missing an init module in its test package.

https://github.com/home-assistant/home-assistant/tree/dev/tests/components/netatmo

If you have time, you can open a separate PR and just add an `__init__.py` module in that directory.

cyberjunky commented:
Done #31189

MartinHjelmare commented on homeassistant/components/garmin_connect/config_flow.py:
```diff
@@ -0,0 +1,75 @@
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        if user_input is None:
+            return await self._show_setup_form(user_input)
```
This method takes errors, not user_input.

MartinHjelmare commented on homeassistant/components/garmin_connect/config_flow.py:
```diff
@@ -0,0 +1,75 @@
+
+        entries = self._async_current_entries()
+        for entry in entries:
+            if entry.data[CONF_ID] == unique_id:
```
Please set a unique id instead. Use `self.async_set_unique_id` and then `self._abort_if_unique_id_configured`.

https://developers.home-assistant.io/docs/en/config_entries_config_flow_handler.html#unique-ids

https://github.com/home-assistant/home-assistant/blob/aa390efd694dea5ea62319621959120218b3ee06/homeassistant/config_entries.py#L819-L853

cyberjunky replied:
You mean replace it by those two calls?
This will result in a 'Aborted' message only, without any explanation why.
Not really an user experience improvement.
How to get this functionality back? (reason="already_setup") 

```

        await self.async_set_unique_id(unique_id)
        self._abort_if_unique_id_configured()

        # entries = self._async_current_entries()
        # for entry in entries:
        #     if entry.data[CONF_ID] == unique_id:
        #         return self.async_abort(reason="already_setup")

```

MartinHjelmare replied:
Fill in the corresponding string with the message.

MartinHjelmare replied:
It's the `"already_configured"` string that we need to set in strings.json.

MartinHjelmare commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,186 @@
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this sensor."""
+        return f"{DOMAIN}_{self._unique_id}_{self._type}"
```
We don't need to include `DOMAIN` in the unique_id. The entity registry is aware of both the integration domain and the platform domain.

MartinHjelmare commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,186 @@
+    def device_info(self) -> Dict[str, Any]:
+        """Return device information."""
+        return {
+            "identifiers": {(DOMAIN, self._unique_id)},
```
We're creating multiple entities but this should just be one device, right? We can't use the entity unique_id as a unique device id.

cyberjunky replied:
I get your point, devices, sensors, entities it becomes confusing, but see this in almost every integration I look at. Should I remove the unique_id part? I should contain at lease DOMAIN, otherwise you get zombie devices not linked to integration.

MartinHjelmare replied:
We can use the unique_id of the config entry, since there's one device per entry.

MartinHjelmare replied:
Sorry, I see now that was what we were using already. I missed the underscore in front so I thought it was the entity unique_id.

cyberjunky replied:
Applied

MartinHjelmare replied:
Thanks! Sorry again.

MartinHjelmare commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,186 @@
+            "model": "Activity Tracker",
+        }
+
+    async def async_added_to_hass(self):
```
If we don't use this, we don't need to implement it.

cyberjunky replied:
It was just there to get rid of the "unknown" state, with a plan to fill it with the watch used, but didn't manage to get something other than "GARMIN" from the portal.
I will remove it.
![Screenshot from 2020-01-27 09-59-35](https://user-images.githubusercontent.com/5447161/73162449-1bbd3400-40ee-11ea-9318-2186944315af.png)

MartinHjelmare commented on homeassistant/components/garmin_connect/sensor.py:
```diff
@@ -0,0 +1,186 @@
+        """Register state update callback."""
+        pass
+
+    async def async_will_remove_from_hass(self):
```
Same as above.

MartinHjelmare commented on tests/components/garmin_connect/__init__py:
```diff
@@ -0,0 +1 @@
+"""Tests for the Garmin Connect component."""
```
There's a missing return at end of file.

cyberjunky committed:
Code improvements from PR feedback

cyberjunky committed:
Fix entity indentifier

cyberjunky committed:
Reverted device identifier

MartinHjelmare commented on homeassistant/components/garmin_connect/strings.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "config": {
+    "abort": {
+      "already_setup": "This account is already configured."
```
```suggestion
      "already_configured": "This account is already configured."
```

cyberjunky committed:
Fixed abort message

MartinHjelmare reviewed (approved):
Looks good!

cyberjunky commented:
I think I have applied all requested changes, I have one thing left; This change let the test_config_flow mock and the check fail again for the already configured test:
```
        await self.async_set_unique_id(unique_id)
        self._abort_if_unique_id_configured()

        # entries = self._async_current_entries()
        # for entry in entries:
        #     if entry.data[CONF_ID] == unique_id:
        #         return self.async_abort(reason="already_configured")
```

cyberjunky committed:
Test fix

MartinHjelmare commented on tests/components/garmin_connect/test_config_flow.py:
```diff
@@ -91,8 +89,9 @@ def mock_garmin():
     """Test abort if already setup."""
-    entry = MockConfigEntry(domain=DOMAIN, data=MOCK_CONF)
-    entry.add_to_hass(hass)
+    result = await hass.config_entries.flow.async_init(
```
Pass the unique_id as a keyword argument to the mock config entry instead.

https://github.com/home-assistant/home-assistant/blob/7d9c8fdfa09323eef42a73dca295ab5fabe1d833/tests/common.py#L656-L672

cyberjunky replied:
Done, thanks again for being super helpful!

cyberjunky committed:
Fixed unique_id MockConfigEntry
